mim_ocr/graph/builders.py

```diff
@@ -0,0 +1,180 @@
+from typing import Tuple
+from abc import ABC, abstractmethod
+from enum import Enum
+
+import numpy as np
+from igraph import Vertex, Graph, Edge
+
+from mim_ocr.data_model.box import Box
+from mim_ocr.graph.graph_utils import boxes_horizontal_distance, boxes_verdical_distance
+
+
+class Builder(ABC):
+    @classmethod
+    def get_box_for_vertex(cls, vertex: Vertex) -> Box:
+        return vertex["box"]
+
+
+class VertexBuilder(Builder):
+    @classmethod
+    @abstractmethod
+    def build(cls, graph: Graph, vertex: Vertex):
+        raise NotImplementedError
+
+
+class EdgeBuilder(Builder):
+    class HorizontalDirection(Enum):
+        LEFT = 0
+        SAME = 1
+        RIGHT = 2
+
+    class VerdicalDirection(Enum):
+        UP = 0
+        SAME = 1
+        DOWN = 2
+
+    @classmethod
+    def build(cls, graph: Graph, vertex1: Vertex):
+        for vertex2 in graph.vs:
+            if vertex1 == vertex2:
+                continue
+            edge_properties = cls._find_properties(vertex1, vertex2)
+            if len(edge_properties["directions"]) == 0:
+                continue
+            # builder_properties = {cls.__name__: edge_properties}
+
+            if graph.are_connected(vertex1, vertex2):
+                edge = graph.es[graph.get_eid(vertex1, vertex2)]
+                edge.update_attributes(**edge_properties)
+            else:
+                graph.add_edge(vertex1, vertex2, **edge_properties)
+
+    @classmethod
+    def is_down_edge(cls, e: Edge):
+        return (
+            "verdical" in e.attributes()["directions"]
+            and cls.HorizontalDirection.DOWN.value
+            is e.attributes()["directions"]["verdical"].value
+        )
+
+    @classmethod
+    def is_up_edge(cls, e: Edge):
+        return (
+            "verdical" in e.attributes()["directions"]
+            and cls.HorizontalDirection.UP.value
+            is e.attributes()["directions"]["verdical"].value
+        )
+
+    @classmethod
+    def is_left_edge(cls, e: Edge):
+        return (
+            "horizontal" in e.attributes()["directions"]
+            and cls.HorizontalDirection.LEFT.value
+            is e.attributes()["directions"]["horizontal"].value
+        )
+
+    @classmethod
+    def is_right_edge(cls, e: Edge):
+        return (
+            "horizontal" in e.attributes()["directions"]
+            and cls.HorizontalDirection.RIGHT.value
+            is e.attributes()["directions"]["horizontal"].value
+        )
+
+    @classmethod
+    @abstractmethod
+    def _find_properties(cls, vertex1: Vertex, vertex2: Vertex):
+        raise NotImplementedError
+
+
+class VerdicalEdgeBuilder(EdgeBuilder):
+    @classmethod
+    def _find_properties(cls, vertex1: Vertex, vertex2: Vertex):
+        box_v = cls.get_box_for_vertex(vertex1)
+        box_u = cls.get_box_for_vertex(vertex2)
```

Comment:
Jeśli już mamy `vertex1, vertex2`, to spodziewałabym się `box1, box2`

Reply:
Jasne, poprawiłem.